File: Game.rb
require_relative 'Labyrinth'
require_relative 'GameCharacter'
require_relative 'GameState'


module Irrgarten
  # Clase que representa el juego Irrgarten.
  #
  # Coordina la creación y configuración del laberinto, los jugadores, los monstruos,
  # el desarrollo de los turnos (movimientos, combate, recompensas o resurrección) y la
  # generación de un log con los eventos del juego.
  #
  #
  # @author Juan Caballero Santoyo
  class Game
    @@MAX_ROUNDS=10

    #Dimensiones del laberinto
    @@ROWS=7
    @@COLS=7


    @@EXIT_ROW
    @@EXIT_COL 

    # Inicializa una nueva instancia del juego.
    #
    # Se crea el laberinto con una salida en una posición aleatoria, se instancian los jugadores,
    # se selecciona el jugador que iniciará el juego y se configuran monstruos y bloques en el laberinto.
    #
    # @param [Integer] nplayers Número de jugadores de la partida.
    # @return [Game] La instancia del juego inicializada.
    def initialize(nplayers)
      @@EXIT_ROW=Dice.random_pos(@@ROWS)
      @@EXIT_COL=Dice.random_pos(@@COLS)

      @players = Array.new
      @monsters = Array.new

      nplayers.times do |i|
        @players.push(Player.new(i.to_s, Dice.random_intelligence, Dice.random_strength))
      end
  
      @current_player_index = Dice.who_starts(nplayers)
      @current_player = @players[@current_player_index]
      
      @log = "-Game just started. \n "
      
      @lab=Labyrinth.new(@@ROWS,@@COLS,@@EXIT_ROW,@@EXIT_COL)
      configure_labyrinth()
      @lab.spread_players(@players)

    end

    # Indica si la partida ha finalizado.
    #
    # Se considera finalizada cuando algún jugador alcanza la celda de salida.
    #
    # @return [Boolean] true si hay un ganador, false en caso contrario.
    def finished
      return @lab.have_a_winner
    end


    # Procesa el siguiente turno del juego.
    #
    # Realiza el movimiento del jugador actual en función de la dirección preferida,
    # gestiona el combate (si se encuentra un monstruo), aplica recompensas o la
    # resurrección en caso de que el jugador esté muerto, y finalmente avanza al siguiente jugador.
    #
    # @param [Direction] preferred_direction Dirección preferida para mover al jugador.
    # @return [Boolean] true si el juego ha finalizado (hay ganador), false en caso contrario.
    def next_step(preferred_direction)
      @log=""

      dead=@current_player.dead

      if(!dead)

        direction=actual_direction(preferred_direction)

        if(direction!= preferred_direction)
          log_player_no_orders()
        end

        monster=@lab.put_player(direction,@current_player)

        if(monster==nil)
          log_no_monster()
        else
          winner=combat(monster)
          manage_reward(winner)
        end
      else
        manage_resurrection()
      end
      
      end_game=finished()

      if(!end_game)
        next_player()
      end

      return end_game
    
    end

    # Retorna el estado actual del juego.
    #
    # Se recopila la representación en cadena del laberinto, la información de cada jugador,
    # la información de los monstruos, el índice del jugador actual, si el juego ha finalizado y el log de eventos.
    #
    # @return [GameState] Objeto que encapsula el estado actual del juego.
    def game_state
      info_players=""
      info_monsters=""

      @players.each do |player|
        info_players+=player.to_s+"\n"
      end

      @monsters.each do |monster|
        info_monsters+=monster.to_s+"\n"
      end

      return GameState.new(@lab.to_s,info_players,info_monsters,@current_player_index, finished(),@log)
    end

    private

    # Configura el laberinto: añade monstruos y bloques en posiciones predefinidas.
    #
    # Se crean tres monstruos, se colocan en el laberinto y se guardan en el array @monsters.
    # También se añaden bloques en determinadas posiciones y orientaciones.
    #
    # @return [void]
    def configure_labyrinth

      monster_1=Monster.new("Mike Wasowsky",Dice.random_intelligence,Dice.random_strength)
      monster_2=Monster.new("Alucard",Dice.random_intelligence,Dice.random_strength)
      monster_3=Monster.new("Frankenstein",Dice.random_intelligence,Dice.random_strength)

      @lab.add_monster(3,3,monster_1)
      @lab.add_monster(1,5,monster_2)
      @lab.add_monster(5,5,monster_3)
      
      @monsters.push(monster_1)
      @monsters.push(monster_2)
      @monsters.push(monster_3)

      @lab.add_block(Orientation::HORIZONTAL,6,3,1)
      @lab.add_block(Orientation::HORIZONTAL,2,1,2)
      @lab.add_block(Orientation::VERTICAL,4,5,2)


    end

    # Avanza al siguiente jugador en la lista.
    #
    # Actualiza el índice y el jugador actual de forma circular.
    #
    # @return [void]
    def next_player
      if(@current_player_index==@players.size-1)
        @current_player_index=0
      else
        @current_player_index+=1
      end

      @current_player=@players[@current_player_index]
    end

    # Calcula la dirección efectiva a partir de la preferida.
    #
    # Consulta al laberinto por los movimientos válidos en la posición actual del jugador
    # y utiliza el método move del jugador para determinar la dirección final.
    #
    # @param [Direction] preferred_direction Dirección preferida por el jugador.
    # @return [Symbol] La dirección efectivamente elegida.
    def actual_direction(preferred_direction)
      current_row=@current_player.row
      current_col=@current_player.col

      valid_moves=@lab.valid_moves(current_row,current_col)

      output=@current_player.move(preferred_direction,valid_moves)

      return output
    end


    # Simula un combate entre el jugador actual y un monstruo.
    #
    # Alterna ataques entre el jugador y el monstruo hasta que uno pierda o se alcance
    # el número máximo de rondas. Se retorna quién resultó vencedor.
    #
    # @param [Monster] monster Monstruo contra el que se combate.
    # @return [GameCharacter] Constante que indica el vencedor (p.ej., GameCharacter::PLAYER o GameCharacter::MONSTER).
    def combat(monster)
      rounds=0
      winner=GameCharacter::PLAYER

      player_attack=@current_player.attack

      lose=monster.defend(player_attack)

      while( (!lose) && (rounds < @@MAX_ROUNDS))
        winner=GameCharacter::MONSTER
        rounds+=1

        monster_attack=monster.attack
        lose=@current_player.defend(monster_attack)

        if(!lose)
          player_attack=@current_player.attack
          winner = GameCharacter::PLAYER
          lose=monster.defend(player_attack)
        end
      end

      log_rounds(rounds, @@MAX_ROUNDS)

      return winner
    end

    # Gestiona la recompensa tras el combate.
    #
    # Si el jugador vence, recibe una recompensa llamando a receive_reward;
    # en caso contrario, se registra la victoria del monstruo.
    #
    # @param [GameCharacter] winner Constante que indica al vencedor del combate.
    # @return [void]
    def manage_reward(winner)
      if(winner==GameCharacter::PLAYER)
        @current_player.receive_reward
        log_player_won()
      else
        log_monster_won()
      end
    end

    # Gestiona la resurrección del jugador actual.
    #
    # Consulta a Dice si el jugador debe resucitar. Si es así, invoca el método de resurrección;
    # si no, registra que el jugador omite su turno.
    #
    # @return [void]
    def manage_resurrection
      resurrect=Dice.resurrect_player
      if(resurrect)
        @current_player.resurrect
        log_resurrected()
      else
        log_player_skip_turn()
      end

    end


    # Registra en el log que el jugador ganó el combate.
    #
    # @return [void]
    def log_player_won
        @log += "- Player #{@current_player_index} won the fight.\n"
    end

    # Registra en el log que el monstruo ganó el combate.
    #
    # @return [void]
    def log_monster_won
        @log += "- Monster won the fight.\n"
    end

    # Registra en el log que el jugador fue resucitado.
    #
    # @return [void]
    def log_resurrected
        @log += "- Player #{@current_player_index} resurrected.\n"
    end

    # Registra en el log que el jugador perdió su turno (estaba muerto).
    #
    # @return [void]
    def log_player_skip_turn
        @log += "- Player #{@current_player_index} skipped turn (is dead).\n"
    end

    # Registra en el log que el jugador no siguió las órdenes (la dirección preferida no era válida).
    #
    # @return [void]
    def log_player_no_orders
        @log += "- Player #{@current_player_index} didn't follow orders, it was not possible.\n"
    end

    # Registra en el log que el jugador se movió a una celda vacía o que el movimiento no fue posible.
    #
    # @return [void]
    def log_no_monster
        @log += "- Player #{@current_player_index} moved to an empty cell or it was not possible to move.\n"
    end

    # Registra en el log la cantidad de rondas transcurridas durante el combate.
    #
    # @param [Integer] rounds Número de rondas efectivamente ejecutadas.
    # @param [Integer] max Número máximo de rondas permitidas.
    # @return [void]
    def log_rounds(rounds, max)
        @log += "- Rounds: #{rounds}/#{max}.\n"
    end
  end
  
end

File: java/src/irrgarten/Game.java

package irrgarten;
import java.util.ArrayList;


/**
 * Clase que representa el juego Irrgarten. Esta clase se encarga de gestionar el estado del juego,
 * las acciones de los jugadores y los monstruos, y de generar un estado del juego que pueda ser
 * consultado por la interfaz de usuario.
 * 
 * @author Juan Caballero Santoyo
 */
public class Game {
    private static final int MAX_ROUNDS=10;
  
            
    private int currentPlayerIndex;
    private String log;
    
    private Labyrinth lab;
    private ArrayList<Monster> monsters;
    private ArrayList<Player> players;
    private Player currentPlayer;
    
    //Dimensiones del laberinto
    private static final int ROWS=7;
    private static final int COLS=7;
    
    
    /**
     * Constructor de la clase Game, que inicializará los jugadores a jugar 
     * (definiendo el jugador que inicia la partida) y el estado del laberinto, 
     * que incluye su propia construcción, es decir, definir los bloques,
     * los monstruos y sus posiciones.
     * 
     * @param nPlayers Numero de jugadores
     */
    public Game (int nPlayers){
        
        int exitRow=Dice.randomPos(ROWS);
        int exitCol=Dice.randomPos(COLS);

        this.players=new ArrayList<>();
        this.monsters=new ArrayList<>();
        
        for(int n = 0; n < nPlayers; n++){
            players.add(new Player(Character.forDigit(n, 10), Dice.randomIntelligence(), Dice.randomStrength()));
        }
                
        this.currentPlayerIndex=Dice.whoStarts(nPlayers);
        this.currentPlayer=this.players.get(this.currentPlayerIndex);
        
        this.lab= new Labyrinth(ROWS, COLS, exitRow, exitCol);
        this.configureLabyrinth();

        this.lab.spreadPlayers(this.players);
        
        this.log="- Game just started.\n";
    }
    
    /**
     * Indica si algún jugador ha ganado la partida, es decir, si ha finalizado
     * el juego
     * @return True si se finalizo el juego, false en caso contrario
     */
    public boolean finished(){
        return lab.haveAWinner();
    }
    
    
    /**
     * Método que estudia un turno completo del juego. Gestiona el paso de un turno a otro.
     * @param preferredDirection  Dirección a la que se pretende mover el jugador
     * @return True si se finalizo el juego, false en caso contrario
     */
    public boolean nextStep(Directions preferredDirection){
        log="";
        boolean dead=currentPlayer.dead();
        
        if(!dead){
            Directions direction=actualDirection(preferredDirection);
            
            if(direction != preferredDirection){
                logPlayerNoOrders();
            }
            
            Monster monster=lab.putPlayer(direction, currentPlayer);
            
            if(monster==null){
                logNoMonster();
            }
            else{
                GameCharacter winner=combat(monster);
                manageReward(winner);
            }
            
        }
        else{
            manageResurrection();
        }
        boolean endGame=finished();
        
        if(!endGame)
            nextPlayer();
        return endGame;
    }
    
    /**
     * Crea una intancia de la clase GameState a partir de los miembros de 
     * la clase Game
     * 
     * @return Una intancia de GameState con los datos del estado actual del juego
     */
    public GameState getGameState(){
       
        String infoPlayers="";
        String infoMonsters="";

        for (int i=0; i<this.players.size(); i++){
            infoPlayers+="- " + this.players.get(i).toString()+"\n";
        }
        
        
        for (int i=0; i<this.monsters.size(); i++){
            infoMonsters+="- " + this.monsters.get(i).toString()+"\n";
        }
        
        GameState estadoGeneral = new GameState (this.lab.toString(), infoPlayers,
                                infoMonsters, this.currentPlayerIndex, this.finished(), this.log);
        
        return estadoGeneral;
    }  
    
    /**
     * Se crea el laberinto a partir de los monstruos y bloques definidos
     */
    private void configureLabyrinth(){
        
        Monster monster_1=new Monster("Mike Wasowsky", Dice.randomIntelligence(),Dice.randomStrength());
        Monster monster_2=new Monster("Alucard", Dice.randomIntelligence(),Dice.randomStrength());
        Monster monster_3=new Monster("Frankenstein", Dice.randomIntelligence(),Dice.randomStrength());
        lab.addMonster(3, 3, monster_1);
        lab.addMonster(1, 5, monster_2);
        lab.addMonster(5, 5, monster_3);
        monsters.add(monster_1);
        monsters.add(monster_2);
        monsters.add(monster_3);
        lab.addBlock(Orientation.HORIZONTAL, 6, 3, 1);
        lab.addBlock(Orientation.HORIZONTAL, 2, 1, 2);
        lab.addBlock(Orientation.VERTICAL, 4, 5, 2);

        
    }
    
    
    
    /**
     * Se indica que el turno pasa al siguiente jugador
     */
    private void nextPlayer(){
        if(currentPlayerIndex==players.size()-1)
            currentPlayerIndex=0;
        else
            ++currentPlayerIndex;
        
        currentPlayer=players.get(currentPlayerIndex);
    }
    
    
    /**
     * Método que mueve de hecho al jugador. Se intenta que se mueva a la casilla
     * indicada por la dirección, y si no es posible, se moverá en otra dirección.
     * 
     * @param preferredDirection  Dirección a la que se pretende mover el jugador
     * @return   Dirección a la que se moverá el jugador (si es válida)
     */
    private Directions actualDirection(Directions preferredDirection){
        
        int currentRow=currentPlayer.getRow();
        int currentCol=currentPlayer.getCol();
        
        ArrayList<Directions> validMoves=lab.validMoves(currentRow,currentCol);

        Directions output=currentPlayer.move(preferredDirection,validMoves);
        
        return output;

    }
    
    /**
     * Método encargado de llevar a cabo un combate entre el jugador actual y un monstruo.
     * 
     * @param monster Monstruo con el que se va a combatir.
     * @return GameCharacter: Jugador o Monstruo, tipo de rol del ganador del combate.
     */
    private GameCharacter combat(Monster monster){
        
         int rounds=0;   
        
        GameCharacter winner=GameCharacter.PLAYER;
        boolean lose = monster.defend(currentPlayer.attack());
        
        while (!lose && rounds<MAX_ROUNDS){ 
            rounds++;   
            winner = GameCharacter.MONSTER;
            lose = currentPlayer.defend(monster.attack());
            
            if (!lose){ 
                winner = GameCharacter.PLAYER;
                lose = monster.defend(currentPlayer.attack());
            }
        }
        this.logRounds(rounds, MAX_ROUNDS);
        return winner;
        
    }
    
    /**
     * Método que gestiona las recompensas que se producen al finalizar un combate.
     * Se encarga de actualizar el log con la información de si ha ganado el jugador
     * o el monstruo.
     * 
     * @param winner Tipo de personaje que ha ganado el combate, puede ser un jugador o un monstruo.
     * @see GameCharacter
     */
    private void manageReward(GameCharacter winner){
        if(winner==GameCharacter.PLAYER){
            currentPlayer.receiveReward();
            logPlayerWon();
        }
        else
            logMonsterWon();
            
    }
    /**
     * Método que gestiona la resurrección de un jugador al finalizar un combate.
     */
    private void manageResurrection(){
        boolean resurrect=Dice.resurrectPlayer();
        
        if(resurrect){
            currentPlayer.resurrect();
            logResurrected();
            
            FuzzyPlayer fuzzy=new FuzzyPlayer(currentPlayer);
            this.players.set(currentPlayerIndex, fuzzy);
            lab.switchToFuzzy(fuzzy);
        }
        else{
            logPlayerSkipTurn();
        }
    }
    
    /**
     * Se actualiza el log con la información de que el jugador actual
     * ha ganado el combate
     */
    private void logPlayerWon(){
        this.log+= "- Player "+this.currentPlayerIndex+" won the fight.\n";
    }
    
    /**
     * Se actualiza el log con la información de que el mosntruo
     * ha ganado el combate al jugador actual
     */
    private void logMonsterWon(){
        this.log+= "- Monster won the fight.\n";
    }
    
    /**
     * Se actualiza el log con la información de que el jugador actual
     * ha ganado resucitado
     */
    private void logResurrected(){
        this.log+= "- Player "+this.currentPlayerIndex+" resurrected as Fuzzy.\n";
    }
    
    /**
     * Se actualiza el log con la información de que el jugador actual
     * ha perdido su turno por estar muerto
     */
    private void logPlayerSkipTurn(){
        this.log+= "- Player "+this.currentPlayerIndex+" skipped turn (is dead).\n";
    }
    
    /**
     * Se actualiza el log con la información de que el jugador actual
     * ha intentado ejecutar alguna acción no permitida y no se ha producido ningún
     * cambio
     */
    private void logPlayerNoOrders(){
        this.log+= "- Player "+this.currentPlayerIndex+" didn't follow orders, it was not possible.\n";
    }
    
    /**
     * Se actualiza el log con la información de que el jugador actual
     * se ha desplazado a una casilla vacía o no ha sido posible el desplazamiento
     */
    private void logNoMonster(){
        this.log+= "- Player "+this.currentPlayerIndex+" moved to an empty square or it was not possible to move.\n";
    }
    
    /**
     * Se actualiza el log con la información de que se han producido rounds
     * de max rounds en el combate actual
     * 
     * @param rounds Número de rounds que se han producido
     * @param max Número máximo de rounds que puede durar un combate
     */
    private void logRounds(int rounds, int max){
        this.log+= "- Rounds: "+rounds+"/"+max+".\n";
    }
    
}